// ts-mls/dist/src/credential.js
import { uint16Decoder, uint16Encoder } from "./codec/number.js";
import { flatMapDecoder, mapDecoder } from "./codec/tlsDecoder.js";
import { contramapBufferEncoders } from "./codec/tlsEncoder.js";
import { varLenDataDecoder, varLenTypeDecoder, varLenDataEncoder, varLenTypeEncoder } from "./codec/variableLength.js";
import { defaultCredentialTypes, isDefaultCredentialTypeValue } from "./defaultCredentialType.js";
/** @public */
export function isDefaultCredential(c) {
    return isDefaultCredentialTypeValue(c.credentialType);
}
export const credentialBasicEncoder = contramapBufferEncoders([uint16Encoder, varLenDataEncoder], (c) => [c.credentialType, c.identity]);
export const credentialX509Encoder = contramapBufferEncoders([uint16Encoder, varLenTypeEncoder(varLenDataEncoder)], (c) => [c.credentialType, c.certificates]);
export const credentialCustomEncoder = contramapBufferEncoders([uint16Encoder, varLenDataEncoder], (c) => [c.credentialType, c.data]);
export const credentialEncoder = (c) => {
    if (!isDefaultCredential(c))
        return credentialCustomEncoder(c);
    switch (c.credentialType) {
        case defaultCredentialTypes.basic:
            return credentialBasicEncoder(c);
        case defaultCredentialTypes.x509:
            return credentialX509Encoder(c);
    }
};
const credentialBasicDecoder = mapDecoder(varLenDataDecoder, (identity) => ({
    credentialType: defaultCredentialTypes.basic,
    identity,
}));
const credentialX509Decoder = mapDecoder(varLenTypeDecoder(varLenDataDecoder), (certificates) => ({ credentialType: defaultCredentialTypes.x509, certificates }));
function credentialCustomDecoder(credentialType) {
    return mapDecoder(varLenDataDecoder, (data) => ({ credentialType, data }));
}
export const credentialDecoder = flatMapDecoder(uint16Decoder, (credentialType) => {
    switch (credentialType) {
        case defaultCredentialTypes.basic:
            return credentialBasicDecoder;
        case defaultCredentialTypes.x509:
            return credentialX509Decoder;
        default:
            return credentialCustomDecoder(credentialType);
    }
});
