// ts-mls/dist/src/leafNode.js
import { capabilitiesEncoder, capabilitiesDecoder } from "./capabilities.js";
import { uint32Decoder, uint32Encoder } from "./codec/number.js";
import { mapDecoders, flatMapDecoder, mapDecoderOption, mapDecoder } from "./codec/tlsDecoder.js";
import { contramapBufferEncoders, encode } from "./codec/tlsEncoder.js";
import { varLenDataEncoder, varLenDataDecoder, varLenTypeEncoder, varLenTypeDecoder } from "./codec/variableLength.js";
import { credentialEncoder, credentialDecoder } from "./credential.js";
import { signWithLabel, verifyWithLabel } from "./crypto/signature.js";
import { extensionEncoder, leafNodeExtensionDecoder } from "./extension.js";
import { leafNodeSources, leafNodeSourceValueDecoder, leafNodeSourceValueEncoder } from "./leafNodeSource.js";
import { lifetimeEncoder, lifetimeDecoder } from "./lifetime.js";
import { fastEqual } from "./util/byteArray.js";
export const leafNodeDataEncoder = contramapBufferEncoders([varLenDataEncoder, varLenDataEncoder, credentialEncoder, capabilitiesEncoder], (data) => [data.hpkePublicKey, data.signaturePublicKey, data.credential, data.capabilities]);
export const leafNodeDataDecoder = mapDecoders([varLenDataDecoder, varLenDataDecoder, credentialDecoder, capabilitiesDecoder], (hpkePublicKey, signaturePublicKey, credential, capabilities) => ({
    hpkePublicKey,
    signaturePublicKey,
    credential,
    capabilities,
}));
export const leafNodeInfoKeyPackageEncoder = contramapBufferEncoders([leafNodeSourceValueEncoder, lifetimeEncoder, varLenTypeEncoder(extensionEncoder)], (info) => [leafNodeSources.key_package, info.lifetime, info.extensions]);
export const leafNodeInfoUpdateOmittedEncoder = contramapBufferEncoders([leafNodeSourceValueEncoder, varLenTypeEncoder(extensionEncoder)], (i) => [i.leafNodeSource, i.extensions]);
export const leafNodeInfoCommitOmittedEncoder = contramapBufferEncoders([leafNodeSourceValueEncoder, varLenDataEncoder, varLenTypeEncoder(extensionEncoder)], (info) => [info.leafNodeSource, info.parentHash, info.extensions]);
export const leafNodeInfoOmittedEncoder = (info) => {
    switch (info.leafNodeSource) {
        case leafNodeSources.key_package:
            return leafNodeInfoKeyPackageEncoder(info);
        case leafNodeSources.update:
            return leafNodeInfoUpdateOmittedEncoder(info);
        case leafNodeSources.commit:
            return leafNodeInfoCommitOmittedEncoder(info);
    }
};
export const leafNodeInfoKeyPackageDecoder = mapDecoders([lifetimeDecoder, varLenTypeDecoder(leafNodeExtensionDecoder)], (lifetime, extensions) => ({
    leafNodeSource: leafNodeSources.key_package,
    lifetime,
    extensions,
}));
export const leafNodeInfoUpdateOmittedDecoder = mapDecoder(varLenTypeDecoder(leafNodeExtensionDecoder), (extensions) => ({
    leafNodeSource: leafNodeSources.update,
    extensions,
}));
export const leafNodeInfoCommitOmittedDecoder = mapDecoders([varLenDataDecoder, varLenTypeDecoder(leafNodeExtensionDecoder)], (parentHash, extensions) => ({
    leafNodeSource: leafNodeSources.commit,
    parentHash,
    extensions,
}));
export const leafNodeInfoOmittedDecoder = flatMapDecoder(leafNodeSourceValueDecoder, (leafNodeSource) => {
    switch (leafNodeSource) {
        case leafNodeSources.key_package:
            return leafNodeInfoKeyPackageDecoder;
        case leafNodeSources.update:
            return leafNodeInfoUpdateOmittedDecoder;
        case leafNodeSources.commit:
            return leafNodeInfoCommitOmittedDecoder;
    }
});
export const leafNodeInfoUpdateEncoder = contramapBufferEncoders([leafNodeInfoUpdateOmittedEncoder, varLenDataEncoder, uint32Encoder], (i) => [i, i.groupId, i.leafIndex]);
export const leafNodeInfoCommitEncoder = contramapBufferEncoders([leafNodeInfoCommitOmittedEncoder, varLenDataEncoder, uint32Encoder], (info) => [info, info.groupId, info.leafIndex]);
export const leafNodeInfoEncoder = (info) => {
    switch (info.leafNodeSource) {
        case leafNodeSources.key_package:
            return leafNodeInfoKeyPackageEncoder(info);
        case leafNodeSources.update:
            return leafNodeInfoUpdateEncoder(info);
        case leafNodeSources.commit:
            return leafNodeInfoCommitEncoder(info);
    }
};
export const leafNodeInfoUpdateDecoder = mapDecoders([leafNodeInfoUpdateOmittedDecoder, varLenDataDecoder, uint32Decoder], (ln, groupId, leafIndex) => ({
    ...ln,
    groupId,
    leafIndex,
}));
export const leafNodeInfoCommitDecoder = mapDecoders([leafNodeInfoCommitOmittedDecoder, varLenDataDecoder, uint32Decoder], (ln, groupId, leafIndex) => ({
    ...ln,
    groupId,
    leafIndex,
}));
export const leafNodeTBSEncoder = contramapBufferEncoders([leafNodeDataEncoder, leafNodeInfoEncoder], (tbs) => [tbs, tbs]);
export const leafNodeEncoder = contramapBufferEncoders([leafNodeDataEncoder, leafNodeInfoOmittedEncoder, varLenDataEncoder], (leafNode) => [leafNode, leafNode, leafNode.signature]);
export const leafNodeDecoder = mapDecoders([leafNodeDataDecoder, leafNodeInfoOmittedDecoder, varLenDataDecoder], (data, info, signature) => ({
    ...data,
    ...info,
    signature,
}));
export const leafNodeKeyPackageDecoder = mapDecoderOption(leafNodeDecoder, (ln) => ln.leafNodeSource === leafNodeSources.key_package ? ln : undefined);
export const leafNodeCommitDecoder = mapDecoderOption(leafNodeDecoder, (ln) => ln.leafNodeSource === leafNodeSources.commit ? ln : undefined);
export const leafNodeUpdateDecoder = mapDecoderOption(leafNodeDecoder, (ln) => ln.leafNodeSource === leafNodeSources.update ? ln : undefined);
function toTbs(leafNode, groupId, leafIndex) {
    switch (leafNode.leafNodeSource) {
        case leafNodeSources.key_package:
            return { ...leafNode, leafNodeSource: leafNode.leafNodeSource };
        case leafNodeSources.update:
            return { ...leafNode, leafNodeSource: leafNode.leafNodeSource, groupId, leafIndex };
        case leafNodeSources.commit:
            return { ...leafNode, leafNodeSource: leafNode.leafNodeSource, groupId, leafIndex };
    }
}
export async function signLeafNodeCommit(tbs, signaturePrivateKey, sig) {
    return {
        ...tbs,
        signature: await signWithLabel(signaturePrivateKey, "LeafNodeTBS", encode(leafNodeTBSEncoder, tbs), sig),
    };
}
export async function signLeafNodeKeyPackage(tbs, signaturePrivateKey, sig) {
    return {
        ...tbs,
        signature: await signWithLabel(signaturePrivateKey, "LeafNodeTBS", encode(leafNodeTBSEncoder, tbs), sig),
    };
}
export function verifyLeafNodeSignature(leaf, groupId, leafIndex, sig) {
    return verifyWithLabel(leaf.signaturePublicKey, "LeafNodeTBS", encode(leafNodeTBSEncoder, toTbs(leaf, groupId, leafIndex)), leaf.signature, sig);
}
export function verifyLeafNodeSignatureKeyPackage(leaf, sig) {
    return verifyWithLabel(leaf.signaturePublicKey, "LeafNodeTBS", encode(leafNodeTBSEncoder, leaf), leaf.signature, sig);
}
export function leafNodeEqual(a, b) {
    return fastEqual(a.signaturePublicKey, b.signaturePublicKey);
}
