// ts-mls/dist/src/message.js
import { flatMapDecoder, mapDecoder } from "./codec/tlsDecoder.js";
import { contramapBufferEncoders } from "./codec/tlsEncoder.js";
import { groupInfoDecoder, groupInfoEncoder } from "./groupInfo.js";
import { keyPackageDecoder, keyPackageEncoder } from "./keyPackage.js";
import { privateMessageDecoder, privateMessageEncoder } from "./privateMessage.js";
import { protocolVersionDecoder, protocolVersionEncoder } from "./protocolVersion.js";
import { publicMessageDecoder, publicMessageEncoder } from "./publicMessage.js";
import { welcomeDecoder, welcomeEncoder } from "./welcome.js";
import { wireformatDecoder, wireformatEncoder, wireformats } from "./wireformat.js";
export const mlsPublicMessageEncoder = contramapBufferEncoders([protocolVersionEncoder, wireformatEncoder, publicMessageEncoder], (msg) => [msg.version, msg.wireformat, msg.publicMessage]);
export const mlsWelcomeEncoder = contramapBufferEncoders([protocolVersionEncoder, wireformatEncoder, welcomeEncoder], (wm) => [wm.version, wm.wireformat, wm.welcome]);
export const mlsPrivateMessageEncoder = contramapBufferEncoders([protocolVersionEncoder, wireformatEncoder, privateMessageEncoder], (pm) => [pm.version, pm.wireformat, pm.privateMessage]);
export const mlsGroupInfoEncoder = contramapBufferEncoders([protocolVersionEncoder, wireformatEncoder, groupInfoEncoder], (gi) => [gi.version, gi.wireformat, gi.groupInfo]);
export const mlsKeyPackageEncoder = contramapBufferEncoders([protocolVersionEncoder, wireformatEncoder, keyPackageEncoder], (kp) => [kp.version, kp.wireformat, kp.keyPackage]);
/** @public */
export const mlsMessageEncoder = (mc) => {
    switch (mc.wireformat) {
        case wireformats.mls_public_message:
            return mlsPublicMessageEncoder(mc);
        case wireformats.mls_welcome:
            return mlsWelcomeEncoder(mc);
        case wireformats.mls_private_message:
            return mlsPrivateMessageEncoder(mc);
        case wireformats.mls_group_info:
            return mlsGroupInfoEncoder(mc);
        case wireformats.mls_key_package:
            return mlsKeyPackageEncoder(mc);
    }
};
/** @public */
export const mlsMessageDecoder = flatMapDecoder(protocolVersionDecoder, (version) => flatMapDecoder(wireformatDecoder, (wireformat) => {
    switch (wireformat) {
        case wireformats.mls_public_message:
            return mapDecoder(publicMessageDecoder, (publicMessage) => ({ version, wireformat, publicMessage }));
        case wireformats.mls_welcome:
            return mapDecoder(welcomeDecoder, (welcome) => ({ version, wireformat, welcome }));
        case wireformats.mls_private_message:
            return mapDecoder(privateMessageDecoder, (privateMessage) => ({ version, wireformat, privateMessage }));
        case wireformats.mls_group_info:
            return mapDecoder(groupInfoDecoder, (groupInfo) => ({ version, wireformat, groupInfo }));
        case wireformats.mls_key_package:
            return mapDecoder(keyPackageDecoder, (keyPackage) => ({ version, wireformat, keyPackage }));
    }
}));
