// ts-mls/dist/src/proposal.js
import { uint16Decoder, uint32Decoder, uint16Encoder, uint32Encoder } from "./codec/number.js";
import { flatMapDecoder, mapDecoder, mapDecoders, orDecoder } from "./codec/tlsDecoder.js";
import { contramapBufferEncoder, contramapBufferEncoders } from "./codec/tlsEncoder.js";
import { varLenDataDecoder, varLenTypeDecoder, varLenDataEncoder, varLenTypeEncoder } from "./codec/variableLength.js";
import { extensionEncoder, groupContextExtensionDecoder } from "./extension.js";
import { keyPackageDecoder, keyPackageEncoder } from "./keyPackage.js";
import { pskIdDecoder, pskIdEncoder } from "./presharedkey.js";
import { decodeDefaultProposalTypeValue, defaultProposalTypeValueEncoder, defaultProposalTypes, isDefaultProposalTypeValue, } from "./defaultProposalType.js";
import { protocolVersionDecoder, protocolVersionEncoder } from "./protocolVersion.js";
import { leafNodeUpdateDecoder, leafNodeEncoder } from "./leafNode.js";
export const addEncoder = contramapBufferEncoder(keyPackageEncoder, (a) => a.keyPackage);
export const addDecoder = mapDecoder(keyPackageDecoder, (keyPackage) => ({ keyPackage }));
export const updateEncoder = contramapBufferEncoder(leafNodeEncoder, (u) => u.leafNode);
export const updateDecoder = mapDecoder(leafNodeUpdateDecoder, (leafNode) => ({ leafNode }));
export const removeEncoder = contramapBufferEncoder(uint32Encoder, (r) => r.removed);
export const removeDecoder = mapDecoder(uint32Decoder, (removed) => ({ removed }));
export const pskEncoder = contramapBufferEncoder(pskIdEncoder, (p) => p.preSharedKeyId);
export const pskDecoder = mapDecoder(pskIdDecoder, (preSharedKeyId) => ({ preSharedKeyId }));
export const reinitEncoder = contramapBufferEncoders([varLenDataEncoder, protocolVersionEncoder, uint16Encoder, varLenTypeEncoder(extensionEncoder)], (r) => [r.groupId, r.version, r.cipherSuite, r.extensions]);
export const reinitDecoder = mapDecoders([varLenDataDecoder, protocolVersionDecoder, uint16Decoder, varLenTypeDecoder(groupContextExtensionDecoder)], (groupId, version, cipherSuite, extensions) => ({ groupId, version, cipherSuite, extensions }));
export const externalInitEncoder = contramapBufferEncoder(varLenDataEncoder, (e) => e.kemOutput);
export const externalInitDecoder = mapDecoder(varLenDataDecoder, (kemOutput) => ({ kemOutput }));
export const groupContextExtensionsEncoder = contramapBufferEncoder(varLenTypeEncoder(extensionEncoder), (g) => g.extensions);
export const groupContextExtensionsDecoder = mapDecoder(varLenTypeDecoder(groupContextExtensionDecoder), (extensions) => ({ extensions }));
/** @public */
export function isDefaultProposal(p) {
    return isDefaultProposalTypeValue(p.proposalType);
}
export const proposalAddEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, addEncoder], (p) => [p.proposalType, p.add]);
export const proposalUpdateEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, updateEncoder], (p) => [p.proposalType, p.update]);
export const proposalRemoveEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, removeEncoder], (p) => [p.proposalType, p.remove]);
export const proposalPSKEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, pskEncoder], (p) => [p.proposalType, p.psk]);
export const proposalReinitEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, reinitEncoder], (p) => [p.proposalType, p.reinit]);
export const proposalExternalInitEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, externalInitEncoder], (p) => [p.proposalType, p.externalInit]);
export const proposalGroupContextExtensionsEncoder = contramapBufferEncoders([defaultProposalTypeValueEncoder, groupContextExtensionsEncoder], (p) => [p.proposalType, p.groupContextExtensions]);
export const proposalCustomEncoder = contramapBufferEncoders([uint16Encoder, varLenDataEncoder], (p) => [p.proposalType, p.proposalData]);
export const proposalEncoder = (p) => {
    if (!isDefaultProposal(p))
        return proposalCustomEncoder(p);
    switch (p.proposalType) {
        case defaultProposalTypes.add:
            return proposalAddEncoder(p);
        case defaultProposalTypes.update:
            return proposalUpdateEncoder(p);
        case defaultProposalTypes.remove:
            return proposalRemoveEncoder(p);
        case defaultProposalTypes.psk:
            return proposalPSKEncoder(p);
        case defaultProposalTypes.reinit:
            return proposalReinitEncoder(p);
        case defaultProposalTypes.external_init:
            return proposalExternalInitEncoder(p);
        case defaultProposalTypes.group_context_extensions:
            return proposalGroupContextExtensionsEncoder(p);
    }
};
export const proposalAddDecoder = mapDecoder(addDecoder, (add) => ({
    proposalType: defaultProposalTypes.add,
    add,
}));
export const proposalUpdateDecoder = mapDecoder(updateDecoder, (update) => ({
    proposalType: defaultProposalTypes.update,
    update,
}));
export const proposalRemoveDecoder = mapDecoder(removeDecoder, (remove) => ({
    proposalType: defaultProposalTypes.remove,
    remove,
}));
export const proposalPSKDecoder = mapDecoder(pskDecoder, (psk) => ({
    proposalType: defaultProposalTypes.psk,
    psk,
}));
export const proposalReinitDecoder = mapDecoder(reinitDecoder, (reinit) => ({
    proposalType: defaultProposalTypes.reinit,
    reinit,
}));
export const proposalExternalInitDecoder = mapDecoder(externalInitDecoder, (externalInit) => ({ proposalType: defaultProposalTypes.external_init, externalInit }));
export const proposalGroupContextExtensionsDecoder = mapDecoder(groupContextExtensionsDecoder, (groupContextExtensions) => ({
    proposalType: defaultProposalTypes.group_context_extensions,
    groupContextExtensions,
}));
export function proposalCustomDecoder(proposalType) {
    return mapDecoder(varLenDataDecoder, (proposalData) => ({ proposalType, proposalData }));
}
export const proposalDecoder = orDecoder(flatMapDecoder(decodeDefaultProposalTypeValue, (proposalType) => {
    switch (proposalType) {
        case defaultProposalTypes.add:
            return proposalAddDecoder;
        case defaultProposalTypes.update:
            return proposalUpdateDecoder;
        case defaultProposalTypes.remove:
            return proposalRemoveDecoder;
        case defaultProposalTypes.psk:
            return proposalPSKDecoder;
        case defaultProposalTypes.reinit:
            return proposalReinitDecoder;
        case defaultProposalTypes.external_init:
            return proposalExternalInitDecoder;
        case defaultProposalTypes.group_context_extensions:
            return proposalGroupContextExtensionsDecoder;
    }
}), flatMapDecoder(uint16Decoder, (n) => proposalCustomDecoder(n)));
