// ts-mls/dist/src/proposalOrRefType.js
import { uint8Decoder, uint8Encoder } from "./codec/number.js";
import { flatMapDecoder, mapDecoder, mapDecoderOption } from "./codec/tlsDecoder.js";
import { contramapBufferEncoders } from "./codec/tlsEncoder.js";
import { varLenDataDecoder, varLenDataEncoder } from "./codec/variableLength.js";
import { proposalDecoder, proposalEncoder } from "./proposal.js";
import { numberToEnum } from "./util/enumHelpers.js";
/** @public */
export const proposalOrRefTypes = {
    proposal: 1,
    reference: 2,
};
export const proposalOrRefTypeEncoder = uint8Encoder;
export const proposalOrRefTypeDecoder = mapDecoderOption(uint8Decoder, numberToEnum(proposalOrRefTypes));
export const proposalOrRefProposalEncoder = contramapBufferEncoders([proposalOrRefTypeEncoder, proposalEncoder], (p) => [p.proposalOrRefType, p.proposal]);
export const proposalOrRefProposalRefEncoder = contramapBufferEncoders([proposalOrRefTypeEncoder, varLenDataEncoder], (r) => [r.proposalOrRefType, r.reference]);
export const proposalOrRefEncoder = (input) => {
    switch (input.proposalOrRefType) {
        case proposalOrRefTypes.proposal:
            return proposalOrRefProposalEncoder(input);
        case proposalOrRefTypes.reference:
            return proposalOrRefProposalRefEncoder(input);
    }
};
export const proposalOrRefDecoder = flatMapDecoder(proposalOrRefTypeDecoder, (proposalOrRefType) => {
    switch (proposalOrRefType) {
        case proposalOrRefTypes.proposal:
            return mapDecoder(proposalDecoder, (proposal) => ({ proposalOrRefType, proposal }));
        case proposalOrRefTypes.reference:
            return mapDecoder(varLenDataDecoder, (reference) => ({ proposalOrRefType, reference }));
    }
});
